mlflow/utils/rest_utils.py

```diff
@@ -201,6 +201,8 @@ def http_request(
             f" To increase the timeout, set the environment variable {MLFLOW_HTTP_REQUEST_TIMEOUT}"
             " to a larger value."
         )
+    except requests.exceptions.InvalidURL as iu:
+        raise InvalidUrlException(f"Invalid artifact url: {url}\n{iu}")
```

Comment:
This function is a generic wrapper for REST requests. If an InvalidURL is passed, the message here doesn't make sense for the vast majority of requests. 
This specific error catch should be localized with specificity to the point of calling (we want to handle invalid URI exception catching directly for artifact uri's with an informative message where we're passing artifact uri's). 
Can this be moved within the cli `gc()` logic? We also want to include the `error_code` with custom exceptions to provide informative error code mapping within the REST handlers.

Reply:
Hi @BenWilson2 

The `http_request`  function now wraps every type of exception as `MlflowException`, and the outer scope cannot recognize the original error type. This causes the cli `gc()` cannot simply classifying the URL error from others using an exception handling process. 
To address this, I have separated the `requests.exceptions.InvalidURL` from the general `Exception` and wrapped it in my custom class called `InvalidUrlException`. This class inherits from `MlflowException` to maintain backward compatibility. Other parts of the code that handle `MlflowException` will cache the `InvalidUrlException`. Prior to this PR, people  would receive an error message like `API request to {url} failed with exception {e_<here the generic exception is my ui>_}` error message from `MlflowException`. However, after this PR, they will receive `Invalid artifact url: {url}\n{iu}` from `InvalidUrlException`.
This way, I can effectively catch this specific URL exception in the gc() process. 
In the case of `requests.exceptions.InvalidURL` occurs when there is no response and error code available. Therefore, I think this exception is independent of the `RestException` processing flow.

Please let me know if there are any corrections needed or if I have misunderstood anything.

Many Thanks and Best Regards,

Reply:
Can some tests be added to confirm this behavior (for both the gc command and other requests that are provided invalid uri's?

Reply:
Sure, please give me some time to find out and prepare the cases.

Reply:
Hi @BenWilson2 

I have edited the first command of this PR and added the experiment results regarding the behavior of calling `http_request()` from both gc and other CLI functions. Would you mind if I resolve this conversation? We can discuss further in the next round of review.